--- a/drivers/rknpu/rknpu_devfreq.c
+++ b/drivers/rknpu/rknpu_devfreq.c
@@ -120,6 +120,51 @@
 		else if (value == 0xa)
 			*bin = 2;
 	}
+	if (of_property_match_string(np, "nvmem-cell-names", "customer_demand") >= 0) {
+		ret = rockchip_nvmem_cell_read_u8(np, "customer_demand", &value);
+		if (ret) {
+			dev_err(dev, "Failed to get customer_demand\n");
+			return ret;
+		}
+		if (value == 0x3)
+			*bin = 4;
+	}
+	if (*bin < 0)
+		*bin = 0;
+	LOG_DEV_INFO(dev, "bin=%d\n", *bin);
+
+	return ret;
+}
+
+static int rv1126b_npu_get_soc_info(struct device *dev, struct device_node *np,
+				    int *bin, int *process)
+{
+	int ret = 0;
+	u8 value = 0;
+
+	if (!bin)
+		return 0;
+
+	if (of_property_match_string(np, "nvmem-cell-names",
+				     "specification_serial_number") >= 0) {
+		ret = rockchip_nvmem_cell_read_u8(
+			np, "specification_serial_number", &value);
+		if (ret) {
+			LOG_DEV_ERROR(
+				dev,
+				"Failed to get specification_serial_number\n");
+			return ret;
+		}
+		/* RV1126BM */
+		if (value == 0xd)
+			*bin = 1;
+		/* RV1126BJ */
+		else if (value == 0xa)
+			*bin = 2;
+		/* RV1126BP */
+		else if (value == 0x10)
+			*bin = 3;
+	}
 	if (*bin < 0)
 		*bin = 0;
 	LOG_DEV_INFO(dev, "bin=%d\n", *bin);
@@ -234,7 +279,7 @@
 
 static const struct rockchip_opp_data rk3576_npu_opp_data = {
 	.set_read_margin = rk3576_npu_set_read_margin,
-	.set_soc_info = rockchip_opp_set_low_length,
+	.set_soc_info = NULL,
 #if KERNEL_VERSION(6, 1, 0) <= LINUX_VERSION_CODE
 	.config_regulators = npu_opp_config_regulators,
 	.config_clks = npu_opp_config_clks,
@@ -251,15 +296,32 @@
 #endif
 };
 
+static const struct rockchip_opp_data rv1126b_npu_opp_data = {
+	.get_soc_info = rv1126b_npu_get_soc_info,
+	.set_soc_info = rk3588_npu_set_soc_info,
+#if KERNEL_VERSION(6, 1, 0) <= LINUX_VERSION_CODE
+	.config_regulators = npu_opp_config_regulators,
+	.config_clks = npu_opp_config_clks,
+#endif
+};
+
 static const struct of_device_id rockchip_npu_of_match[] = {
 	{
 		.compatible = "rockchip,rk3576",
 		.data = (void *)&rk3576_npu_opp_data,
 	},
 	{
+		.compatible = "rockchip,rk3576s",
+		.data = (void *)&rk3576_npu_opp_data,
+	},
+	{
 		.compatible = "rockchip,rk3588",
 		.data = (void *)&rk3588_npu_opp_data,
 	},
+	{
+		.compatible = "rockchip,rv1126b",
+		.data = (void *)&rv1126b_npu_opp_data,
+	},
 	{},
 };
 
@@ -411,7 +473,7 @@
 	struct rknpu_device *rknpu_dev = dev_get_drvdata(dev);
 	struct rockchip_opp_info *opp_info = &rknpu_dev->opp_info;
 
-	if (opp_info->is_scmi_clk) {
+	if (false) {
 		if (clk_set_rate(opp_info->clk, POWER_DOWN_FREQ))
 			LOG_DEV_ERROR(dev, "failed to restore clk rate\n");
 	}
@@ -439,7 +501,7 @@
 	if (opp_info->data && opp_info->data->set_read_margin)
 		opp_info->data->set_read_margin(dev, opp_info,
 						opp_info->target_rm);
-	if (opp_info->is_scmi_clk) {
+	if (false) {
 		if (clk_set_rate(opp_info->clk, rknpu_dev->rate))
 			LOG_DEV_ERROR(dev, "failed to restore clk rate\n");
 	}
